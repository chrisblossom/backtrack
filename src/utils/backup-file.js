/* @flow */

import path from 'path';
import { existsSync } from 'fs';
import { move } from 'fs-extra';
import { getFileHash } from './get-file-hash';

import { rootPath } from '../config/paths';

type Return = Promise<?{
    file: string,
    filename: string,
}>;

async function backupFile(file: string): Return {
    const hash = getFileHash(file);

    const { base: filename, dir, ext = '' } = path.parse(file);

    const shortHash = hash.slice(0, 8);

    let backupFilename = `${filename}-backup-${shortHash}${ext}`;
    let backupFilePath = path.resolve(dir, backupFilename);

    /**
     * Handle the same file being backed up multiple times
     */
    let previouslyBackedUp = false;
    let backupFileExists = existsSync(backupFilePath);
    if (backupFileExists) {
        let count = 0;
        const maxCount = 10;
        while (
            backupFileExists &&
            previouslyBackedUp === false &&
            count < maxCount
        ) {
            const backupFileHash = getFileHash(backupFilePath);

            /**
             * Check if the existing file has the same hash.
             *
             * Prevents backing up the same file over
             */
            if (backupFileHash === hash) {
                previouslyBackedUp = true;
            }

            backupFilename = `${filename}-backup-${shortHash}-${count}${ext}`;
            backupFilePath = path.resolve(dir, backupFilename);

            backupFileExists = existsSync(backupFilePath);
            count += 1;
        }

        /**
         * Prevent flooding filesystem with backups
         */
        if (count === maxCount && backupFileExists) {
            const relativeFile = path.relative(rootPath, file);

            throw new Error(`Remove all existing backups of: ${relativeFile}`);
        }
    }

    /**
     * Don't backup anything if the file has already been backed up
     */
    if (previouslyBackedUp === false) {
        /**
         * Will fail if backup file already exists
         */
        await move(file, backupFilePath, {
            overwrite: false,
        });

        return {
            file: backupFilePath,
            filename: backupFilename,
        };
    }

    return undefined;
}

export { backupFile };
