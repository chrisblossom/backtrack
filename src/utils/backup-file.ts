import path from 'path';
import { existsSync } from 'fs';
import { move } from 'fs-extra';
import { rootPath } from '../config/paths';
import { getFileHash } from './get-file-hash';

type Return = Promise<
	| {
			file: string;
			filename: string;
	  }
	| undefined
>;

async function backupFile(file: string): Return {
	const hash = getFileHash(file);

	const { base: filename, dir, ext = '' } = path.parse(file);

	const shortHash = hash.slice(0, 8);

	let backupFilename = `${filename}-backup-${shortHash}${ext}`;
	let backupFilePath = path.resolve(dir, backupFilename);

	/**
	 * Handle the same file being backed up multiple times
	 */
	let previouslyBackedUp = false;
	let backupFileExists = existsSync(backupFilePath);
	if (backupFileExists) {
		let count = 0;
		const maxCount = 10;
		while (
			backupFileExists &&
			previouslyBackedUp === false &&
			count < maxCount
		) {
			const backupFileHash = getFileHash(backupFilePath);

			/**
			 * Check if the existing file has the same hash.
			 *
			 * Prevents backing up the same file over
			 */
			if (backupFileHash === hash) {
				previouslyBackedUp = true;
			}

			backupFilename = `${filename}-backup-${shortHash}-${count.toString()}${ext}`;
			backupFilePath = path.resolve(dir, backupFilename);

			backupFileExists = existsSync(backupFilePath);
			count += 1;
		}

		/**
		 * Prevent flooding filesystem with backups
		 */
		if (count === maxCount && backupFileExists) {
			const relativeFile = path.relative(rootPath, file);

			throw new Error(`Remove all existing backups of: ${relativeFile}`);
		}
	}

	/**
	 * Don't backup anything if the file has already been backed up
	 */
	if (previouslyBackedUp === false) {
		/**
		 * Will fail if backup file already exists
		 */
		await move(file, backupFilePath, {
			overwrite: false,
		});

		return {
			file: backupFilePath,
			filename: backupFilename,
		};
	}

	return undefined;
}

export { backupFile };
