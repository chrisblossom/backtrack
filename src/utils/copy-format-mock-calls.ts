/**
 * copy.mock.calls is async so order is not guaranteed for expected results
 * Converts to plain object
 *
 * For testing purposes only
 *
 * meant to be used as a snapshot with jest-serializer-path
 */

type CopyCalls = [string, string, {}];

function copyFormatMockCalls(calls: ReadonlyArray<CopyCalls>) {
	const result = calls.reduce((acc, call) => {
		const src = call[0];
		const dest = call[1];
		const options = call[2];

		return {
			...acc,
			[src]: {
				dest,
				options,
			},
		};
	}, {});

	return result;
}

export { copyFormatMockCalls };
