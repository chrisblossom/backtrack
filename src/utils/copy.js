/* @flow */

import fse from 'fs-extra';
import { toArray } from './object-utils';

import type { CopyFile } from '../types.js';

/**
 * See https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md#copysrc-dest-options-callback
 * for options
 */
async function copy(files: $ReadOnlyArray<CopyFile> | CopyFile, options: {}) {
    const normalized = toArray(files);

    /**
     * Does not return any information
     */
    await Promise.all(
        normalized.map((file) => {
            return fse.copy(file.src, file.dest, options);
        }),
    );
}

export { copy };
