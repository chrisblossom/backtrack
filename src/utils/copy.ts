import path from 'path';
import fs from 'fs';
import fse from 'fs-extra';
import { readDirDeep } from 'read-dir-deep';
import { getFileHash } from './get-file-hash';
import { toArray } from './object-utils';

function stat(pathname: string) {
	return new Promise((resolve, reject) => {
		fs.stat(pathname, (error, stats) => {
			if (error) {
				reject(error);

				return;
			}

			resolve(stats);
		});
	});
}

export type File = Readonly<{
	src: string;
	dest: string;

	hash?: boolean;

	/**
	 * fs-extra copy options
	 * https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md#copysrc-dest-options-callback
	 */
	overwrite?: boolean;
	errorOnExist?: boolean;
	preserveTimestamps?: boolean;
}>;

function getHashedName(source: string, dest: string) {
	const parsed = path.parse(dest);
	const fileHash = getFileHash(source).substring(0, 8);

	const rename = `${parsed.name}.${fileHash}${parsed.ext}`;

	const result = path.join(parsed.dir, rename);

	return result;
}

/**
 * See https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md#copysrc-dest-options-callback
 * for options
 */
function copy(files: ReadonlyArray<File> | File): Promise<void> {
	const normalized = toArray(files);

	const copyFilesResult = normalized.map(async (file) => {
		const {
			src,
			hash = false,

			overwrite = true,
			errorOnExist = true,
			preserveTimestamps = true,
		} = file;

		const fseCopyOptions = {
			overwrite,
			errorOnExist,
			preserveTimestamps,
		};

		let dest = file.dest;

		// @ts-ignore
		const isDirectory: Boolean = (await stat(src)).isDirectory();
		if (isDirectory) {
			const deepFileList: string[] = await readDirDeep(src);

			type CopyFilesToDestAcc = {
				src: string;
				dest: string;
				overwrite: boolean;
				errorOnExist: boolean;
				preserveTimestamps: boolean;
			};

			const copyFilesToDest = deepFileList.reduce(
				(acc: CopyFilesToDestAcc[], sourceFile) => {
					const absoluteSourcePath = path.resolve(src, sourceFile);

					let absoluteDestinationPath = path.resolve(
						dest,
						sourceFile,
					);

					if (hash === true) {
						absoluteDestinationPath = getHashedName(
							absoluteSourcePath,
							absoluteDestinationPath,
						);
					}

					return [
						...acc,
						{
							src: absoluteSourcePath,
							dest: absoluteDestinationPath,
							...fseCopyOptions,
						},
					];
				},
				[],
			);

			/**
			 * copy all directory files
			 */
			return copy(copyFilesToDest);
		}

		if (hash === true) {
			dest = getHashedName(src, file.dest);
		}

		return fse.copy(src, dest, fseCopyOptions);
	});

	return Promise.all(copyFilesResult).then(() => {
		/**
		 * fse.copy does not return any values. explicitly return undefined for promise.all
		 */
		return undefined;
	});
}

export { copy };
