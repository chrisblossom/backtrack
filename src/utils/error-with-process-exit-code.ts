/**
 * Adds exitCode to Error for process.exit(exitCode);
 *
 * https://nodejs.org/api/process.html#processexitcode
 */
class ErrorWithProcessExitCode extends Error {
	public exitCode?: number;

	constructor(message: string, exitCode: number) {
		super(message);

		this.exitCode = exitCode;
		// Set the prototype explicitly for better compatibility across different environments
		Object.setPrototypeOf(this, ErrorWithProcessExitCode.prototype);
	}
}

export { ErrorWithProcessExitCode };
