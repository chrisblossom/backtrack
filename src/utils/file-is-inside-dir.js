/* @flow */

import path from 'path';
import { realpathSync } from 'fs';
import { rootPath } from '../config/paths';

/**
 * get the real file path to handle symlinks
 *
 * Not used to check if file exists. Return original file if realpathSync fails
 *
 * WARNING: NOT PROPERLY TESTED
 */
const getRealPath = (file) => {
    try {
        return realpathSync(file);
    } catch (error) {
        return file;
    }
};

/**
 * Check if a file is inside a directory. Handles absolute and relative
 * Assumes rootPath if not specified
 */
function fileIsInsideDir(file: string, dir: string = rootPath): boolean {
    if (!file) {
        return false;
    }

    const normalizedDir = path.normalize(dir) + path.sep;
    const absoluteFilePath = path.resolve(normalizedDir, file);

    const realPath = getRealPath(absoluteFilePath);

    const truncatedPath = realPath.slice(0, normalizedDir.length);

    return truncatedPath === normalizedDir;
}

export { fileIsInsideDir };
