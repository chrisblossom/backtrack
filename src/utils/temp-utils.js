/* @flow */

import path from 'path';
import os from 'os';
import del from 'del';
import {
    readFileSync,
    realpathSync,
    writeFileSync,
} from 'fs';
import { createHash } from 'crypto';
import fse from 'fs-extra';
import parentModule from 'parent-module';
import readPkgUp from 'read-pkg-up';
import { readDirDeepSync } from './read-dir-deep';

class TempUtils {
    dir: string;

    constructor() {
        const tempDir = realpathSync(os.tmpdir());
        const parent = parentModule();
        const relativeParent = path.relative(process.cwd(), parent);

        /**
         * create base directory based on package name
         */
        const baseDir = readPkgUp
            .sync({
                cwd: parent,
                normalize: false,
            })
            .pkg.name // replace special characters for directory name
            .replace(/[^a-zA-Z0-9]/g, '-');

        /**
         * Each temp directory will be unique to the file
         */
        this.dir = path.resolve(tempDir, baseDir, `${relativeParent}-dir`);

        // create directory
        fse.ensureDirSync(this.dir);

        (this: any).absolutePath = this.absolutePath.bind(this);
        (this: any).createDir = this.createDir.bind(this);
        (this: any).createFile = this.createFile.bind(this);
        (this: any).deleteFile = this.deleteFile.bind(this);
        (this: any).readFile = this.readFile.bind(this);
        (this: any).getAllFiles = this.getAllFiles.bind(this);
        (this: any).getFileHash = this.getFileHash.bind(this);
        (this: any).getAllFilesHash = this.getAllFilesHash.bind(this);
        (this: any).clean = this.clean.bind(this);
        (this: any).deleteTempDir = this.deleteTempDir.bind(this);
    }

    absolutePath(dir: string) {
        return path.join(this.dir, dir);
    }

    createDir(dir: string) {
        const dirname = this.absolutePath(dir);
        fse.ensureDirSync(dirname);
    }

    createFile(file: string, contents: string | Object) {
        const filePath = this.absolutePath(file);
        const fileDir = path.parse(filePath).dir;
        fse.ensureDirSync(fileDir);

        if (typeof contents === 'object') {
            // eslint-disable-next-line no-param-reassign
            contents = JSON.stringify(contents);
        }

        // all files append new line
        if (typeof contents === 'string' && contents.slice(-1) !== '\n') {
            // eslint-disable-next-line no-param-reassign
            contents += '\n';
        }

        writeFileSync(filePath, `${contents}`);
    }

    deleteFile(file: string) {
        const filePath = this.absolutePath(file);

        const removed = del.sync(filePath, { root: this.dir });

        return removed;
    }

    readFile(file: string) {
        const filePath = this.absolutePath(file);

        let result = readFileSync(filePath, 'utf8');
        try {
            result = JSON.parse(result);
            // eslint-disable-next-line no-empty
        } catch (e) {}

        return result;
    }

    getFileHash(file: string) {
        const filePath = this.absolutePath(file);

        const fileBody = readFileSync(filePath);
        const hash = createHash('md5')
            .update(fileBody)
            .digest('hex');

        return hash;
    }

    getAllFiles() {
        const fileList = readDirDeepSync(this.dir);

        return fileList;
    }

    getAllFilesHash() {
        const fileList = this.getAllFiles();
        const result = fileList.reduce((acc, file) => {
            const fileHash = this.getFileHash(file);
            return {
                ...acc,
                [file]: fileHash,
            };
        }, {});

        return result;
    }

    clean() {
        const cleanPattern = this.absolutePath('**/*');
        del.sync(cleanPattern, {
            root: this.dir,
            dot: true,
        });
    }

    deleteTempDir() {
        del.sync(this.dir, { force: true, dot: true });
    }
}

export { TempUtils };
