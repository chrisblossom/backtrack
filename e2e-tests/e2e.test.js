/* @flow */

import { TempUtils } from '../src/utils/temp-utils';

const backtrack = () => require('../src/cli/start').start();

const temp = new TempUtils();
const cwd = process.cwd();
const runMode = process.env.RUN_MODE;
const handleErrorMock = jest.fn();
let processExitSpy;

beforeEach(() => {
    process.chdir(temp.dir);
    temp.clean();

    // Suppress all console logging
    jest.spyOn(console, 'info').mockImplementation(() => undefined);
    jest.spyOn(console, 'warn').mockImplementation(() => undefined);
    jest.spyOn(console, 'error').mockImplementation(() => undefined);
    jest.spyOn(console, 'debug').mockImplementation(() => undefined);

    processExitSpy = jest
        .spyOn(process, 'exit')
        .mockImplementation(() => undefined);
    jest.doMock('../src/utils/handle-error.js', () => ({
        handleError: (...args) => {
            handleErrorMock(...args);

            // $FlowIgnore
            const handleError = require.requireActual(
                '../src/utils/handle-error.js',
            ).handleError;

            return handleError(...args);
        },
    }));
});

afterEach(() => {
    process.chdir(cwd);
    process.env.RUN_MODE = runMode;
    jest.restoreAllMocks();
});

afterAll(() => {
    temp.deleteTempDir();
});

function getError() {
    const result = handleErrorMock.mock.calls[0][0];

    if (!result.error) {
        return result;
    }

    if (result.error.message) {
        result.message = result.error.message;
    }

    if (result.error.code) {
        result.code = result.error.code;
    }

    delete result.error;

    return result;
}

test('exits when no config found', async () => {
    process.env.RUN_MODE = 'init';

    await backtrack();

    expect(processExitSpy).toHaveBeenCalledWith(1);

    const error = getError();
    expect(error.message).toEqual('backtrack config not found');
});

test('backtrack', async () => {
    process.env.RUN_MODE = 'init';

    temp.createFile('package.json', { name: 'test-package' });
    temp.createFile('files/file1.js', '// file1.js');
    temp.createFile('static/static-file.js', '// static-file.js');
    temp.createFile('static-1/static1-file.js', '// static1-file.js');

    const config = {
        clean: {
            del: '**/*',
            copy: [
                {
                    src: 'static',
                    dest: 'static',
                },
                {
                    src: 'static-1/static1-file.js',
                    dest: 'static/static1-file.js',
                },
            ],
        },
        files: {
            src: 'files/file1.js',
            dest: 'file1.js',
        },
        packageJson: {
            scripts: {
                test: 'jest',
            },
        },
        dev: ['echo'],
    };

    temp.createFile(
        'backtrack.config.js',
        `module.exports = ${JSON.stringify(config)}`,
    );

    /**
     * Initial
     */
    await backtrack();

    expect(temp.readFile('package.json')).toEqual({
        name: 'test-package',
        scripts: {
            dev: 'backtrack dev --development',
            clean: 'backtrack clean',
            test: 'jest',
        },
    });

    const initialFileHash = temp.getAllFilesHash();
    expect(initialFileHash).toMatchSnapshot();

    expect(processExitSpy).toHaveBeenCalledTimes(0);
    expect(handleErrorMock).toHaveBeenCalledTimes(0);

    /**
     * Ensure no changes after initialization
     */
    await backtrack();
    expect(temp.getAllFilesHash()).toEqual(initialFileHash);

    /**
     * runs dev script
     */
    jest.resetModules();
    process.env.RUN_MODE = 'dev';
    await backtrack();

    /**
     * Removes dev and static dir
     */
    jest.resetModules();
    process.env.RUN_MODE = 'init';

    delete config.dev;
    delete config.clean.copy;

    temp.createFile(
        'backtrack.config.js',
        `module.exports = ${JSON.stringify(config)}`,
    );

    await backtrack();

    expect(temp.readFile('package.json')).toEqual({
        name: 'test-package',
        scripts: {
            clean: 'backtrack clean',
            test: 'jest',
        },
    });

    expect(temp.getAllFilesHash()).toMatchSnapshot();

    /**
     * Removes everything
     */
    jest.resetModules();
    process.env.RUN_MODE = 'init';

    temp.createFile('backtrack.config.js', 'module.exports = {}');

    await backtrack();

    expect(temp.readFile('package.json')).toEqual({ name: 'test-package' });

    expect(temp.getAllFilesHash()).toMatchSnapshot();
});
