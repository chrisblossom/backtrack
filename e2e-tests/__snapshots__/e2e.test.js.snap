// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backtrack 1`] = `
Object {
  ".backtrack-stats.json": "561ad38ed44501267c8cbe42db1796fc",
  "backtrack.config.js": "31457a97787120a9f36fcb742f897502",
  "file1.js": "7f477fcd51e87d9e65b134b17771dc03",
  "files/file1.js": "7f477fcd51e87d9e65b134b17771dc03",
  "package.json": "d3cd662a2371fd1e79441bee77bdc0ac",
}
`;

exports[`backtrack 2`] = `
Object {
  ".backtrack-stats.json": "58836dacb73ced285211ec08f2bf384a",
  "backtrack.config.js": "285892e0eeb9483d6bdb973f3c9c9855",
  "file1.js": "7f477fcd51e87d9e65b134b17771dc03",
  "files/file1.js": "7f477fcd51e87d9e65b134b17771dc03",
  "package.json": "8017eeefdcdaa6e07153a3b5042537e8",
}
`;

exports[`backtrack 3`] = `
Object {
  ".backtrack-stats.json": "6b6c16e551598a31afb4cdb307c906a4",
  "backtrack.config.js": "73c0f77ffd19a4043aca19410be35b8c",
  "files/file1.js": "7f477fcd51e87d9e65b134b17771dc03",
  "package.json": "e5ace72ad475dc0d19ef2425ea09bdd1",
}
`;
